File: frontend/src/model/entities/Story.ts
import type {UUIDTypes} from "uuid";
import type {Priority} from "../enums/Priority.ts";
import type {WorkingState} from "../enums/WorkingState.ts";
import {Task} from "./Task.ts";

export class Story {
    id: UUIDTypes;
    projectId: UUIDTypes;
    userId: UUIDTypes;
    name: string;
    description: string;
    priority: Priority;
    createdAt: Date;
    state: WorkingState;
    tasks: Array<Task>;

    constructor(
        id: UUIDTypes,
        name: string,
        description: string,
        priority: Priority,
        state: WorkingState,
        userId: UUIDTypes,
        projectId: UUIDTypes,
        tasks: Task[]
    ) {
        this.id = id;
        this.createdAt = new Date();
        this.name = name;
        this.description = description;
        this.priority = priority;
        this.state = state;
        this.userId = userId;
        this.projectId = projectId;
        this.tasks = tasks;
    }

    static fromJSON(obj: any): Story {
        const tasks = Array.isArray(obj.tasks)
            ? obj.tasks.map(Task.fromJSON)
            : [];

        return new Story(
            obj.id as UUIDTypes,
            obj.name,
            obj.description,
            obj.priority as Priority,
            obj.state as WorkingState,
            obj.userId as UUIDTypes,
            obj.projectId as UUIDTypes,
            tasks);
    }

    updateDescription(newDescription: string) {
        this.description = newDescription;
    }

    updateState(newState: WorkingState) {
        this.state = newState;
    }

    addTask(task: Task) {
        this.tasks.push(task);
    }
}

File: backend/WebApi/Entities/Story.cs
namespace API.Database.Entities;

using global::API.Database.Entities.Base;
using global::API.Database.Entities.Enums;
using global::API.DTO.Requests.Story;
using MongoDB.Bson.Serialization.Attributes;

public class Story : EntityBase<Guid>
{
    [BsonElement("name")]
    public required Guid ProjectId { get; init; }

    [BsonElement("user_id")]
    public required string UserLogin { get; init; }

    [BsonElement("name")]
    public string Name { get; private set; }

    [BsonElement("description")]
    public string Description { get; private set; }

    [BsonElement("priority")]
    public Priority Priority { get; private set; }

    [BsonElement("working_state")]
    public WorkingState State { get; private set; }

    [BsonElement("created_at")]
    public required DateTime CreatedAt { get; init; }

    public void Update(UpdateStoryDto dto)
    {
        this.Name = dto.Name;
        this.Description = dto.Description;
        this.Priority = dto.Priority;
        this.State = dto.State;
    }

    public static Story Create(Guid projectId, string loginId, CreateStoryDto dto)
    {
        return new Story
        {
            Id = Guid.NewGuid(),
            Name = dto.Name,
            Description = dto.Description,
            ProjectId = projectId,
            UserLogin = loginId,
            Priority = dto.Priority,
            State = WorkingState.TODO,
            CreatedAt = DateTime.UtcNow,
        };
    }
}